settings.txt
# Point size of text (the line height in pixels), the number of lines to
# display, the number of characters on each line, the number of seconds between
# cursor blinks or 0 to switch off blinking, and whether to print events and
# actions (on or off).

points = 18
rows = 24
columns = 80
blink = 0.5
testing = off

# -----------------------------------------------------------------------------
# Browser commands to display help files (in snipe/help/). Each is tried until
# one works. Put your preference first. The open, start, xdg-open commands run
# the default browser on macOS, Windows, Linux (but xdg-open is often not set up
# properly, so it may be better to give a specific browser command).

help = open start chrome google-chrome chromium-browser firefox xdg-open

# -----------------------------------------------------------------------------
# List of font names, and the file for each font (in snipe/fonts/). When a
# character is not found in a specified font, the fonts following it in the list
# are searched in turn. The Symbols2 font contains visible pictures of control
# characters. The Noto family has fonts covering all languages. Install more of
# them, or any other ttf fonts, in snipe/fonts then include the names here. The
# fonts need not be monospaced (but Snipe does not currently support combining
# characters, shaping, or bi-directionality).

fonts = mono sans italic symbols
mono = NotoSansMono-Regular.ttf
sans = NotoSans-Regular.ttf
italic = NotoSans-Italic.ttf
symbols = NotoSansSymbols2-Regular.ttf

# -----------------------------------------------------------------------------
# Key mappings from events to actions. For event X, if S_X or C_X is missing, it
# defaults to X, and SC_X defaults to C_X. Save means synchronise, and possibly
# compile and run. Bigger and Smaller change font size. C_= and C_+ should be
# the same for keyboards with + above =. C is CTRL, or CMD on macOS.
# PAGE_UP/PAGE_DOWN/HOME/END are FN_UP/FN_DOWN/FN_LEFT/FN_RIGHT on macOS.

BACKSPACE = CutLeftChar
ENTER     = Newline
LEFT      = MoveLeftChar
RIGHT     = MoveRightChar
UP        = MoveUpLine
DOWN      = MoveDownLine
PAGE_UP   = PageUp
PAGE_DOWN = PageDown
C_O       = Open
C_X       = Cut
C_C       = Copy
C_V       = Paste
C_Q       = Quit
C_H       = Help
C_S       = Save
C_+       = Bigger
C_=       = Bigger
C_-       = Smaller
C_ENTER   = CycleTheme

# These actions are not yet implemented.
# C_BACKSPACE = CutLeftWord
# HOME        = MoveStartLine
# S_HOME      = MarkStartLine
# END         = MoveEndLine
# S_END       = MarkEndLine
# S_LEFT      = MarkLeftChar
# C_LEFT      = MoveLeftWord
# SC_LEFT     = MarkLeftWord
# S_RIGHT     = MarkRightChar
# C_RIGHT     = MoveRightWord
# SC_RIGHT    = MarkRightWord
# S_UP        = MarkUpLine
# S_DOWN      = MarkDownLine
# DELETE      = CutRightChar
# C_DELETE    = CutRightWord
# C_Y         = Redo
# C_Z         = Undo

# -----------------------------------------------------------------------------
# Named colours in hex rgb format. (These are the Solarized colours.)

base03  = #002b36
base02  = #073642
base01  = #586e75
base00  = #657b83
base0   = #839496
base1   = #93a1a1
base2   = #eee8d5
base3   = #fdf6e3
yellow  = #b58900
orange  = #cb4b16
cyan    = #2aa198
blue    = #268bd2
red     = #dc322f
magenta = #d33682
violet  = #6c71c4
green   = #859900

# -----------------------------------------------------------------------------
# Named styles. A style is the name of a foreground colour, the name of a
# background colour, and the name of a font. If the font, or font and background
# colour, are missing, then the ones defined for the cursor in the current theme
# are used.

light = base01 base3 mono
dark  = base1 base03 mono
badL  = base01 red mono
badD  = base1 red mono

# -----------------------------------------------------------------------------
# Themes (Solarized Light and Dark). The list of themes gives the default style
# for each theme. Each row after that defines the highlighting for a token type.
# Each column represents a theme. Each entry is a style name or a colour name.

themes    = light   dark
COMMENTED = base1   base01
ERROR     = badL    badD
FUNCTION  = violet  violet
ID        = base00  base0
KEY       = cyan    cyan
OP        = magenta magenta
PROPERTY  = red     red
QUOTED    = yellow  yellow
RESERVED  = orange  orange
SIGN      = base00  base0
TYPE      = blue    blue
VALUE     = green   green


src/settings.c
// Snipe settings. Free and open source. See licence.txt.
#include "settings.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>

struct settings {
    char *text;
    char **lines;
    char ***variables;
};

static void splitLines(settings *ss) {
    char *text = ss->text;
    int n = 0;
    for (int i = 0; text[i] != '\0'; i++) if (text[i] == '\n') n++;
    ss->lines = malloc((n+1) * sizeof(char *));
    ss->lines[n] = NULL;
    int row = 0;
    char *line = text;
    for (int i = 0; text[i] != '\0'; i++) {
        if (text[i] != '\n') continue;
        text[i] = '\0';
        if (i > 0 && text[i-1] == '\r') text[i-1] = '\0';
        ss->lines[row++] = line;
        line = &text[i+1];
    }
    ss->variables = malloc((n+1) * sizeof(char **));
    ss->variables[n] = NULL;
}

static void splitWords(settings *ss, int r) {
    char *line = ss->lines[r];
    if (line[0] == '#') { ss->variables[r] = NULL; return; }
    int n = 0;
    for (int i = 0; line[i] != '\0'; i++) {
        if (line[i] == ' ') continue;
        if (i == 0 || line[i-1] == ' ') n++;
    }
    if (n == 0) { ss->variables[r] = NULL; return; }
    ss->variables[r] = malloc((n+1) * sizeof(char *));
    ss->variables[r][n] = NULL;
    int w = 0;
    for (int i = 0; line[i] != '\0'; i++) {
        if (line[i] == ' ') { line[i] = '\0'; continue; }
        if (i == 0 || line[i-1] == '\0') ss->variables[r][w++] = &line[i];
    }
    if (n < 2 || strcmp(ss->variables[r][1], "=") != 0) {
        crash("bad setting in settings.txt line %d", r+1);
    }
}

settings *newSettings(files *fs) {
    settings *ss = malloc(sizeof(settings));
    char *filename = join(2, installDir(fs), "settings.txt");
    ss->text = readPath(filename);
    free(filename);
    splitLines(ss);
    for (int r = 0; ss->lines[r] != NULL; r++) {
        splitWords(ss, r);
    }
    return ss;
}

void freeSettings(settings *ss) {
    for (int r = 0; ss->lines[r] != NULL; r++) {
        if (ss->variables[r] != NULL) free(ss->variables[r]);
    }
    free(ss->variables);
    free(ss->lines);
    free(ss->text);
    free(ss);
}

// Get the i'th value of the given variable, or NULL.
char const *getSetting(settings *ss, char const *v, int i) {
    for (int r = 0; ss->lines[r] != NULL; r++) {
        if (ss->variables[r] == NULL) continue;
        if (strcmp(ss->variables[r][0], v) != 0) continue;
        return ss->variables[r][2+i];
    }
    crash("no setting found for %s", v);
    return NULL;
}

#ifdef TESTsettings

int main(int n, char const *args[]) {
    files *fs = newFiles(args[0]);
    settings *ss = newSettings(fs);
    assert(strcmp(getSetting(ss, "points", 0), "18") == 0);
    assert(strcmp(getSetting(ss, "rows", 0), "24") == 0);
    assert(strcmp(getSetting(ss, "columns", 0), "80") == 0);
    assert(strcmp(getSetting(ss, "help", 0), "open") == 0);
    assert(strcmp(getSetting(ss, "help", 1), "start") == 0);
    assert(strcmp(getSetting(ss, "help", 2), "chrome") == 0);
    freeSettings(ss);
    freeFiles(fs);
    printf("Settings module OK\n");
    return 0;
}

#endif
